status-im/nimbus-eth1
# Kurtosis testing for nimbus eth1 and eth2

Fixes #172 
The kurtosis package that will spin up a private Ethereum testnet over Docker or Kubernetes with multi-client support, Flashbot's mev-boost infrastructure for PBS-related testing/validation, and other useful network tools (transaction spammer, monitoring tools, etc). Kurtosis packages are entirely reproducible and composable, so this will work the same way over Docker or Kubernetes, in the cloud or locally on your machine. For more details https://github.com/kurtosis-tech/ethereum-package

This PR adds easier support to spin up test networks and run testnets both locally and over the github CI. For running the testnet and then the checks run the `run-kurtosis-check.sh` script. This will spin up a testnet with your local `nimbus-eth1` build and pull the `nimbus-eth2` build from `statusim` dockerhub, then run the following tests

- Check if atleast one client is ready
- Check if all EL and CL clients are synced
  - Check if CL clients are synced
  - Check if EL clients are synced
- Check consensus finality
- Check consensus attestation stats
- Check consensus reorgs
- Check consensus forks

- Check if all clients propose blocks with legacy EOA transactions
- 10 EOA transactions per block
- check block proposals
- Execute all opcodes as contract deployment 
- Execute all opcodes as contract call
- Check precompile calls
- Check chain stability ( finality checks )

The networks configs can be easily changes and multi-client testing can be introduced in this whole setup only, by just changing the network configs in `kurtosis-network-params.yml`

> NOTE: This kurtosis CI check is now set to fail, when the checks pass
